src/core/lib/security/authorization/audit_logging.cc

```diff
@@ -0,0 +1,101 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/security/authorization/audit_logging.h"
+
+#include <initializer_list>
+#include <map>
+#include <memory>
+#include <utility>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/string_view.h"
+
+#include <grpc/grpc_audit_logging.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/json/json.h"
+
+namespace grpc_core {
+namespace experimental {
+
+void AuditLoggerRegistry::RegisterAuditLoggerFactory(
+    std::unique_ptr<AuditLoggerFactory> factory) {
+  auto& registry = GetAuditLoggerRegistry();
+  MutexLock lock(&registry.mu_);
+  registry.logger_factories_map_[factory->name()] = std::move(factory);
+}
+
+bool AuditLoggerRegistry::AuditLoggerFactoryExists(absl::string_view name) {
+  auto& registry = AuditLoggerRegistry::GetAuditLoggerRegistry();
+  MutexLock lock(&registry.mu_);
+  return registry.logger_factories_map_.find(name) !=
+         registry.logger_factories_map_.end();
+}
+
+absl::StatusOr<std::unique_ptr<AuditLoggerFactory::Config>>
+AuditLoggerRegistry::ParseAuditLoggerConfig(absl::string_view name,
+                                            const Json& json) {
+  auto factory_or = AuditLoggerRegistry::GetAuditLoggerFactory(name);
+  if (!factory_or.ok()) {
+    return absl::InvalidArgumentError("unsupported audit logger type");
+  }
+  return factory_or.value()->ParseAuditLoggerConfig(json);
+}
+
+std::unique_ptr<AuditLogger> AuditLoggerRegistry::CreateAuditLogger(
+    std::unique_ptr<AuditLoggerFactory::Config> config) {
+  auto factory_or = AuditLoggerRegistry::GetAuditLoggerFactory(config->name());
+  GPR_ASSERT(factory_or.ok());
+  return factory_or.value()->CreateAuditLogger(std::move(config));
+}
+
+AuditLoggerRegistry& AuditLoggerRegistry::GetAuditLoggerRegistry() {
+  static AuditLoggerRegistry& registry = *new AuditLoggerRegistry();
+  return registry;
+}
+
+absl::StatusOr<AuditLoggerFactory*> AuditLoggerRegistry::GetAuditLoggerFactory(
+    absl::string_view name) {
+  auto& registry = GetAuditLoggerRegistry();
+  MutexLock lock(&registry.mu_);
+  auto it = registry.logger_factories_map_.find(name);
+  if (it != registry.logger_factories_map_.end()) {
+    return it->second.get();
+  }
+  return absl::NotFoundError(
+      absl::StrFormat("audit logger factory %s does not exist", name));
+}
+
+void AuditLoggerRegistry::TestOnlyResetRegistry() {
+  auto& registry = GetAuditLoggerRegistry();
+  MutexLock lock(&registry.mu_);
+  registry.logger_factories_map_.clear();
```

Comment:
Note that we are going to want to automatically register all of the built-in logger factories (i.e., the stdout logger) in the `AuditLoggerRegistry` ctor, since we won't want applications to need to explicitly register built-in loggers.

I think that means that this needs to be structured a little differently, because we'll want this to reset the registry to its default state, which will include the built-in loggers.  So I think what we want here is something like this:

```
namespace {
Mutex* g_mu = new Mutex();  // Replaces AuditLoggerRegistry::mu_
AuditLoggerRegistry* g_registry ABSL_GUARDED_BY(g_mu) =
    new AuditLoggerRegistry();
}  // namespace

void AuditLoggerRegistry::TestOnlyResetRegistry() {
  MutexLock lock(g_mu);
  delete g_registry;
  g_registry = new AuditLoggerRegistry();
}
```

Reply:
I followed this pattern and restructured all the relevant methods. I also made the ctor public again since this namespace variable needs to call it. Is this pattern using global variables fine? The only alternative I can think of still making it local to the function-static would be to have the `GetAuditLoggerRegistry` function conditionally reset the instance.

Reply:
(Had the response typed for a while but forgot to hit Comment).

I followed your suggested pattern and restructured all the methods.

Reply:
Instead of making the ctor public, how about something like this:

```
namespace {
Mutex* g_mu = new Mutex();
AuditLoggerRegistry* g_registry ABSL_GUARDED_BY(g_mu) = nullptr;
}  // namespace

// Private static method in AuditLoggerRegistry.
AuditLoggerRegistry& AuditLoggerRegistry::Get() {
  MutexLock lock(g_mu);
  if (g_registry == nullptr) g_registry = new AuditLoggerRegistry();
  return g_registry;
}
```

Reply:
With both the mutex and registry as static class members, the ctor can remain private now.